Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,296 @@
+        return;
+      }
+    }
+  }
+
+  bool empty() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.empty();
+  }
+
+  size_t size() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.size();
+  }
+
+ private:
+  struct Entry {
+    bool operator<(const Entry& rhs) const {
+      return due < rhs.due;
+    }
+
+    T value;
+    TimePoint due;
+  };
+  // All removals are on the front, but most insertions are expected to be on
+  // the back.
+  using Container = std::deque<Entry>;
+  using Iterator = typename Container::iterator;
+
+  void InsertPreservingOrder(Entry&& new_entry) {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    const auto insertion_point =
+        std::upper_bound(scheduled_.begin(), scheduled_.end(), new_entry);
+    scheduled_.insert(insertion_point, std::move(new_entry));
+
+    cv_.notify_one();
+  }
+
+  // This function expects the mutex to be already locked.
+  bool HasDue() const {
+    namespace chr = std::chrono;
+    const auto now = chr::time_point_cast<Duration>(chr::system_clock::now());
+    return !scheduled_.empty() && now >= scheduled_.front().due;
+  }
+
+  // This function expects the mutex to be already locked.
+  void DoPop(T* const out, const Iterator where) {
+    FIREBASE_ASSERT_MESSAGE(!scheduled_.empty(),
+                            "Trying to pop an entry from an empty queue.");
+
+    if (out) {
+      *out = std::move(where->value);
+    }
+    scheduled_.erase(where);
+    cv_.notify_one();
+  }
+
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+  Container scheduled_;
+};
+
+class AsyncQueue;
+
+// A non-owning handle to an operation scheduled in the future, allowing to
+// cancel the operation.
+class DelayedOperation {
+ public:
+  // If the operation has not been run yet, cancels the operation. Otherwise,
+  // it's a no-op.
+  void Cancel();
+
+ private:
+  using Id = unsigned int;
+
+  // Don't allow callers to create their own `DelayedOperation`s.
+  friend class AsyncQueue;
+  DelayedOperation(AsyncQueue* const queue, const Id id)
+      : queue_{queue}, id_{id} {
+  }
+
+  AsyncQueue* const queue_ = nullptr;
+  const Id id_ = 0;
+};
+
+// A serial queue that executes provided operations on a dedicated background
+// thread.
+//
+// Operations may be scheduled for immediate or delayed execution. Operations
+// scheduled for the exact same time are FIFO ordered. Immediate operations
+// always come before delayed operations.
+//
+// The operations are executed sequentially; only a single operation is executed
+// at any given time.
+//
+// Delayed operations may be canceled if they have not already been run.
+class AsyncQueue {
+ public:
+  using Operation = std::function<void()>;
+  using Milliseconds = std::chrono::milliseconds;
+
+ private:
+  using TimePoint = Schedule<Operation, Milliseconds>::TimePoint;
+  // To allow canceling operations, each scheduled operation is assigned
```

Comment:
An alternative would be a `shared/weak_ptr` pair, similar to libdispatch-based version. However, it seemed both more complicated and less efficient.